design: per command help text criteria (#4792)
aws/aws-sam-cli: designs/per_command_help_text.md

                                  resources that can affect permissions in
                                  your AWS account, for example, by creating
                                  new AWS Identity and Access Management (IAM)
                                  users. For those stacks, you must explicitly
                                  acknowledge their capabilities by specifying
                                  this parameter. The only valid valuesare
                                  CAPABILITY_IAM and CAPABILITY_NAMED_IAM. If
                                  you have IAM resources, you can specify
                                  either capability. If you have IAM resources
                                  with custom names, you must specify
                                  CAPABILITY_NAMED_IAM. If you don't specify
                                  this parameter, this action returns an
                                  InsufficientCapabilities error.
  -h, --help                      Show this message and exit.

```

Users would probably have a ton of questions based on that help text. Some things that come up top of mind are below.

* Which options are required?

> * What are the options that are AWS infra related? AWS Credentials related?

Reviewer: I really like this. I think we should go one step further and specify if a command itself requires AWS credentials or not. One of the first things in the help text should be something like:
```
AWS credentials ARE required ~ this command interacts with AWS APIs, meaning that valid AWS credentials are required for this command to be run.
```
or
```
AWS credentials ARE NOT required ~ this command does not interact with AWS APIs, and makes changes that are localized to executing machine.
```

* Which ones are about additional information for developers?
* How can options be ordered such that is about incremental exposure to complexity?

What will be changed?
---------------------

The help text could be changed to look like below. Only the options are showcased here.

Some of the help text also needs changing as there are sentences that do not fully compute.

```bash

Options:

>   Required Options:

Reviewer: I really like partitioning the commands into sections like this!

                                          command creates it.  [required]

  Debug Options:

  --debug                         Turn on debug logging to print debug message generated by SAM CLI and display
                                  timestamps.

  Configuration Options:
  Learn more about configuration files at:  https://docs.aws.amazon.com/serverless-application-
                                  model/latest/developerguide/serverless-sam-cli-config.html.
  
  --config-env TEXT               The environment name specifying the default parameter values in the configuration
                                  file to use. [default: "default"]

  --config-file TEXT              The path and file name of the configuration file containing default parameter values
                                  to use. [default:'samconfig.toml']

  AWS Credential Options:
  
  --profile TEXT                  Select a specific profile from your credential file to get AWS credentials.
  --region TEXT                   Set the AWS Region of the service (e.g. us-east-1).

  Infra Options:

  --image-repository              AWS ECR repostiroty URI where referenced image artifaccts are uploaded. 

>   --image-repositories            Mapping of Function Logical ID to AWS ECR Reposiroty URI

Reviewer: ```suggestion
  --image-repositories            Mapping of Function Logical ID to AWS ECR Repository URI
```

                                  Example: Function_Logical_ID=ECR_Repo_Uri.
                                  This option can be specified multiple times.

  --s3-bucket TEXT                The name of the AWS S3 bucket where artifacts referenced in your template are uploaded.

>   --s3-prefix TEXT                A prefix name that is added to the artifacts name when it's uploaded 
>                                   the AWS S3 bucket. The prefix name is a path name (folder name) for the S3 bucket.

Reviewer: ```suggestion
  --s3-prefix TEXT                A prefix name that is added to the artifact's name when the artifact is uploaded 
                                  to the AWS S3 bucket. The prefix name is a path name (folder name) for the S3 bucket.
```

  --kms-key-id TEXT               The ID of an AWS KMS key that the command uses to encrypt artifacts that are at rest
                                  in the S3 bucket.

  --role-arn TEXT                 The Amazon Resource Name (ARN) of an AWS Identity and Access Management (IAM) role
                                  that AWS CloudFormation assumes when executing the change set.

  --parameter-overrides           A string that contains AWS CloudFormation parameter overrides encoded as
                                  key=value pairs.
                                  
                                  For example, 'ParameterKey=KeyPairName,ParameterValue=MyKey
                                  ParameterKey=InstanceType,ParameterValue=t1.micro' or KeyPairName=MyKey
                                  InstanceType=t1.micro

  --metadata                      A map of metadata to attach to ALL the artifacts that are referenced in
                                  your template.

  --notification-arns LIST        Amazon  Simple  Notification  Service  topic.
                                  Amazon  Resource  Names  (ARNs) that AWS CloudFormation associates with the stack.

>   --tags                          A list of tags to associate with the stack that is created or updated.AWS

Reviewer: ```suggestion
  --tags                          A list of tags to associate with the stack that is created or updated. AWS
```

                                  CloudFormation also propagates these tags to resources in the stack if the resource
                                  supports it.

  --capabilities LIST             A list of capabilities that you must specify before AWS Cloudformation can create
                                  certain stacks. Some stack templates might include resources that can affect
                                  permissions in your AWS account, for example, by creating new AWS Identity and
                                  Access Management (IAM) users. For those stacks, you must explicitly acknowledge
                                  their capabilities by specifying this parameter. The only valid values are
                                  CAPABILITY_IAM and CAPABILITY_NAMED_IAM. If you have IAM resources, you can specify
                                  either capability. If you have IAM resources with custom names, you must specify
                                  CAPABILITY_NAMED_IAM. If you don't specify this parameter, this action returns an
                                  InsufficientCapabilities error.
                                  

>   Advanced Options:

Reviewer: I know it's common for tools to have `Advanced` options, but could the word "advanced" potentially dissuade a newer user from venturing down and trying them out. Not sure if this is a real problem or what a better term would be, mainly just thinking out loud.

Author: This is the right train of thought maybe it should just be named "Additional" options.

>   --code                          Sync code resources. This includes AWS Lambda Functions, API Gateway, and Step

Reviewer: ```suggestion
  --code                          Sync ONLY code resources. This includes AWS Lambda Functions, API Gateway, and Step
```

                                  Functions.

  --watch                         Watch local files and automatically sync with remote.
  --resource-id TEXT              Sync code for all the resources with the ID. To sync a resource within a nested
                                  stack, use the following pattern {ChildStack}/{logicalId}.

  --resource RESOURCE
                                  Sync code for all resources of the given resource type. Accepted values are
                                  ['AWS::Serverless::Function', 'AWS::Lambda::Function',
                                  'AWS::Serverless::LayerVersion', 'AWS::Lambda::LayerVersion',
                                  'AWS::Serverless::Api', 'AWS::ApiGateway::RestApi', 'AWS::Serverless::HttpApi',
                                  'AWS::ApiGatewayV2::Api', 'AWS::Serverless::StateMachine',
                                  'AWS::StepFunctions::StateMachine']

  --dependency-layer / --no-dependency-layer

>                                   This option separates the dependencies of individual function into another layer,
>                                   for speeding up the sync.process

Reviewer: ```suggestion
                                  This option separates the dependencies of an individual function into another layer,
                                  used for speeding up the sync process.
```

  -s, --base-dir DIRECTORY        Resolve relative paths to function's source code with respect to this folder. Use
                                  this if SAM template and your source code are not in same enclosing folder. By
                                  default, relative paths are resolved with respect to the SAM template's location

  -u, --use-container             If your functions depend on packages that have natively compiled dependencies, use
                                  this flag to build your function inside an AWS Lambda-like Docker container

                                  
```

Success criteria for the change
-------------------------------

> * The success criteria for the change is less time spent on the help text to understand which options are useful.

Reviewer: I think introducing some colour/sectioning will help with this a lot. Looking at the base help text, the bold and underlined sections make it a lot easier to quickly see what's going on. It directs a user's eyes to where they should be looking, instead of just seeing a massive wall of text.

Author: Definitely. All the required ones should have it. 

I'll try to add the NO_COLOR or --no-color along with this or scope it separately.

    SNS: Simple Notification Service.
    ECR: Elastic Container Registy
    KMS: Key Management Service.

Options:

  Required Options:

  -t, --template-file, --template PATH    AWS SAM template file.  [default: template.[yaml|yml|json]]
  --stack-name TEXT                       Name of the AWS CloudFormation stack. 

  AWS Credential Options:
  
  Exploratory DESIGN NOTE: How can we streamline credentials
  
  --profile TEXT                  Named profile for AWS credentials.
  --region TEXT                   Set the AWS Region. (e.g. us-east-1). 

  Infrastructure Options:

  --parameter-overrides           String that contains AWS CloudFormation parameter overrides encoded as
                                  key=value pairs.
                                  
                                  For example, 'ParameterKey=KeyPairName,ParameterValue=MyKey
                                  ParameterKey=InstanceType,ParameterValue=t1.micro' or KeyPairName=MyKey
                                  InstanceType=t1.micro

  --capabilities LIST             List of capabilities that one must specify before AWS Cloudformation can create
                                  certain stacks.

>                                   More info at: https://docs.aws.amazon.com/serverlessrepo/latest/devguide/acknowledging-application-capabilities.html

Reviewer 2: Would it be worth it to add the list of valid capabilities here so that we can save a click for those that are aware of what this option does?

Author: It was an exercise to just reduce the amount of text this had, adding the valid ones makes sense.

                                  
```

Here's a look at revamped sam build help text.

```commandline
Usage: sam build [OPTIONS] [RESOURCE_LOGICAL_ID]

  Build your AWS serverless function code.

  To use this command, update your AWS SAM template to specify the path
  to your function's source code in the resource's Code or CodeUri property.

  To build on your workstation, run this command in folder containing
  SAM template. Built artifacts will be written to .aws-sam/build folder

  Examples:
  
  $ sam build

  To build inside a AWS Lambda like Docker container
  $ sam build --use-container

  To build with inline environment variables passed inside build containers
  $ sam build --use-container --container-env-var Function.ENV_VAR=value --container-env-var GLOBAL_ENV_VAR=value

>   To build with environment variables file passd inside build containers

Author: I changed passed to within to make it simpler.
